contracts/contracts/Multisig.cairo

```diff
@@ -470,52 +426,131 @@ func execute_transaction{
+end
+
+# Sets the owners field on the multisig. The only way this can be invoked
+# is via a recursive call from execute_transaction -> set_owners.
+# Number of required confirmations is decreased in case its larger than
+# number of new owners.
+@external
+func set_owners{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+    owners_len : felt, owners : felt*
+):
+    require_multisig()
+    _set_owners(owners_len, owners)
+
+    return ()
+end
+
+# Set new owners and number of required confirmations.
+# Can be called only via recursively from
+# execute_transaction -> set_owners_and_confirmations_required
+@external
+func set_owners_and_confirmations_required{
+    syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr
+}(owners_len : felt, owners : felt*, confirmations_required : felt):
+    set_owners(owners_len, owners)
+    set_confirmations_required(confirmations_required)
+
+    return ()
+end
+
 #
 # Storage Helpers
 #
 
-func _set_owners{
-        syscall_ptr : felt*,
-        pedersen_ptr : HashBuiltin*,
-        range_check_ptr,
-    }(
-        owners_index : felt,
-        owners_len : felt,
-        owners : felt*,
-    ):
+func _get_owners{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+    owners_index : felt, owners_len : felt, owners : felt*
+):
+    if owners_index == owners_len:
+        return ()
+    end
+
+    let (owner) = _owners.read(index=owners_index)
+    assert owners[owners_index] = owner
+
+    _get_owners(owners_index=owners_index + 1, owners_len=owners_len, owners=owners)
+    return ()
+end
+
+func _set_owners{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(
+    owners_len : felt, owners : felt*
+):
+    alloc_locals
+    require_unique_owners(owners_len, owners)
+
+    # Clean previous owners
+    let (old_owners_len) = _owners_len.read()
+    _clean_owners_range(0, old_owners_len)
+
+    let (local confirmations_required) = _confirmations_required.read()
+    # checks if owners_len < confirmations_required
+    let (lt) = is_le(owners_len, confirmations_required - 1)
+    if lt == TRUE:
+        # Lower number of confirmation
+        _confirmations_required.write(owners_len)
+        ConfirmationsSet.emit(owners_len)
+
+        tempvar syscall_ptr = syscall_ptr
+        tempvar pedersen_ptr : HashBuiltin* = pedersen_ptr
+        tempvar range_check_ptr = range_check_ptr
+    else:
+        tempvar syscall_ptr = syscall_ptr
+        tempvar pedersen_ptr : HashBuiltin* = pedersen_ptr
+        tempvar range_check_ptr = range_check_ptr
+    end
+    tempvar syscall_ptr = syscall_ptr
+    tempvar pedersen_ptr : HashBuiltin* = pedersen_ptr
+    tempvar range_check_ptr = range_check_ptr
+
+    let (tx_valid_since) = _next_tx_index.read()
+    _tx_valid_since.write(tx_valid_since)
+
+    _owners_len.write(owners_len)
+    # Recursively write the rest
+    _set_owners_range(0, owners_len, owners)
+    OwnersSet.emit(owners_len, owners)
```

Comment:
The order of operations in this function is strange:

1. Clean old owners
2. Set threshold
3. Set new owners

Let's consider changing it to something more expected and consistent with other workflows, so we preserve a consistent order of raised events (owners set first, then threshold).

1. Clean old owners
2. Set new owners
3. Set threshold

Reply:
As discussed in the other comment, let's actually remove the threshold setting from here and move it (check + set) to the top-level `set_owners` function.

Reply:
Consider resolved!